### ISSUE TITLE
Complete sourcing page for importing CSV

### ISSUE THREAD

**Issue by karimbahgat** [wmgeolab/scope#44]
A suggestion here is to turn each row in the csv file into a dict, and then pass that dict as kwargs when creating new Source instances. 

<del>This should maybe also delete all existing sources? Alternatively have a separate button to clear all existing sources?</del>

Later on, we will want to only show this functionality for the admin during that the initial setup stage, but for now this should be fine.

**Comment by micrittenden** [comment]
I don't think the csv function should delete all existing sources. Can imagine that causing absolute mayhem lol.
I agree that the "clear all sources" button should be something separate that only project managers have access to. It could be on a pop-up form which requires user authentication or password or something? Like right next to the import button, there's a "delete" button --> project manager clicks button --> form pops up asking for authentication --> says "This action will delete all sources and is irreversible. Are you sure you want to delete them?" and has a red delete button. Alternatively, this form could take a range of source IDs to delete (could be all, could be one, could be some).
Regardless our decision on the delete button, I think it is a separate issue from the import csv button/function.

**Comment by karimbahgat** [comment]
Good point, the delete-all button should be a separate task issue, and warrants some further discussion. Maybe for the purpose of the prototype, clearing the list of sources has to be done in the MySQL database directly, so that the import function simply adds to any existing sources.

**Comment by micrittenden** [comment]
I agree that's good for the prototype

**Comment by micrittenden** [comment]
Import button should:
- [ ] send user to a new page where they select the local file to upload

New page for importing should:
- [ ] have a browse computer button
- [ ] have something which indicates the file path once selected
- [ ] have a submit button
- [ ] have a cancel button

Submit button should:
- [ ] run the import
- [ ] display the import (either "All successful, or these rows were skipped")
- [ ] render a "Return to Home" button once the import is finished

Cancel button should:
- [ ] return to home screen without saving any changes

The import should:
- [ ] import row by row and indicate any rows which were skipped (ie failed) in the process (duplicate URL, blank fields (which shouldn't be blank), typos in source_code, etc)

"Return to Home" button should:
- [ ] return to home screen, where the newly imported data should be visible

**Comment by micrittenden** [comment]
Possible rework:
change the submit input button to a test button which onclick() runs checks to make sure all the data would load properly and displays the results on the screen. Then, the user can click "Submit Changes" (the actual submit button which sends to the view) or "Cancel"

**Comment by micrittenden** [comment]
Django doesn't allow forms to be carried over multiple urls, so we weren't able to implement the above steps as desired. Instead, we just have the results of the import show on the next page. Maybe we can create an intermediary page where users can check their data. Alternatively, if we saved the file in a temp folder, would we be able to reference it twice then delete it?

**Comment by karimbahgat** [comment]
So there are ways to have multi-page/multi-step forms in Django. However, where the trouble is, is in remember which file was chosen from an html file picker input across multiple pages, which is a limitation at the level of the browser/html (otherwise it would be easy to sneak a little piece of html code that would upload files from your computer without you knowing it). The version we first worked on together did have an intermediary page where the user first tests the file, and then brings you to a screen that allows you to upload -- only problem being, once you get to the upload page you would have to manually browse and find your file again. We could revert to that behavior if you really prefer having the test functionality. There are other ways that include saving to a temporary folder as you suggest, as well as ajax javascript stuff, but that's gonna be a bit of a rabbit hole with many traps along the way. 

I would suggest creating a new issue for improving the CSV import page so that we revisit this later once the prototype is finished.

**Comment by karimbahgat** [comment]
I created a new issue for improving the csv import, so that we can revisit this later.
